src/BinCommand.php

```diff
@@ -33,57 +34,85 @@ protected function configure()
 
         putenv('COMPOSER_BIN_DIR='.$this->createConfig()->get('bin-dir'));
 
-        $binVendorRoot = 'vendor-bin';
+        $binvendorRoot = 'vendor-bin';
         $binNamespace = $input->getArgument('namespace');
         $input = new StringInput(preg_replace('/bin\s+' . preg_quote($binNamespace, '/') . '/', '', $input->__toString(), 1));
 
-        if ('all' !== $binNamespace) {
-            $binRoot = $binVendorRoot.'/'.$binNamespace;
-            if (!file_exists($binRoot)) {
-                mkdir($binRoot, 0777, true);
-            }
-
-            $this->chdir($binRoot);
-
-            return $this->getApplication()->doRun($input, $output);
-        }
+        return ('all' !== $binNamespace)
+            ? $this->executeInNamespace($application, $binvendorRoot.'/'.$binNamespace, $input, $output)
+            : $this->executeAllNamespaces($application, 'vendor-bin', $input, $output)
+        ;
+    }
 
+    /**
+     * @param ComposerApplication $application
+     * @param                     $binVendorRoot
+     * @param InputInterface      $input
+     * @param OutputInterface     $output
+     *
+     * @return int
+     */
+    private function executeAllNamespaces(ComposerApplication $application, $binVendorRoot, InputInterface $input, OutputInterface $output)
+    {
         $binRoots = glob($binVendorRoot.'/*', GLOB_ONLYDIR);
         if (empty($binRoots)) {
             $this->getIO()->writeError('<warning>Couldn\'t find any bin namespace.</warning>');
 
-            return;
+            return 1;
         }
 
         $originalWorkingDir = getcwd();
         $exitCode = 0;
         foreach ($binRoots as $binRoot) {
-            $this->chdir($binRoot);
-
-            $exitCode += $this->getApplication()->doRun($input, $output);
+            $exitCode += $this->executeInNamespace($application, $binRoot, $input, $output);
 
             chdir($originalWorkingDir);
-            $this->resetComposers();
+            $this->resetComposers($application);
         }
 
         return min($exitCode, 255);
     }
 
     /**
+     * @param ComposerApplication $application
+     * @param string              $namespace
+     * @param InputInterface      $input
+     * @param OutputInterface     $output
+     *
+     * @return int
+     */
+    private function executeInNamespace(ComposerApplication $application, $namespace, InputInterface $input, OutputInterface $output)
+    {
+        if (!file_exists($namespace)) {
+            mkdir($namespace, 0777, true);
+        }
+
+        $this->chdir($namespace);
+
+        return $application->doRun($input, $output);
+    }
+
+    /**
      * {@inheritDoc}
      */
     public function isProxyCommand()
     {
         return true;
     }
 
-    private function resetComposers()
+    /**
+     * Resets all Composer references in the application.
+     *
+     * @param ComposerApplication $application
+     */
+    private function resetComposers(ComposerApplication $application)
```

Comment:
There were some issues when changing directories and a composer application is loaded somewhere, IIRC it's mostly due to relative configs which were resolved with outdated full path, this makes sure there is a clean composer before each namespace command (cf. here for instance https://github.com/composer/composer/blob/22100b21eef2bb36fdbb47b23ddb69d2fab75fc3/src/Composer/Config.php#L34).